File: 김금환/힙(Heap)/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

// 최소 힙
priority_queue<int, vector<int>, greater<int>> foods;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    
    for (int food : scoville) {
        foods.push(food);
    }
    
    // 모든 음식의 스코빌 지수가 아직 K 이상이 아닌 동안
    while (foods.top() < K) {
        
        // 모든 음식의 스코빌 지수를 K 이상으로 만들 수 없는 경우
        if (foods.size() < 2) {
            answer = -1;
            break;
        }
        
        // 음식 섞기
        int food1 = foods.top();
        foods.pop();
        int food2 = foods.top();
        foods.pop();
        foods.push(food1 + food2 * 2);
        
        ++answer;
    }
    
    return answer;
}

File: 이예진/힙(Heap)/더맵게.py
import heapq
def solution(scoville, K):
    answer = 0
    # 힙큐 만들기
    heapq.heapify(scoville)
    
    while True:
        # 길이 작으면 런타임에러 뜰까봐
        if len(scoville) < 2:
            break
            
        # 가장 작은 거
        min1 = heapq.heappop(scoville)
        # 근데 가장 작은 게 K보다 크면 나가
        if min1 >= K:
            break
        # 아니라면 두 번째로 작은 거 pop해서 더하고 곱해야함
        else:
            min2 = heapq.heappop(scoville)
            heapq.heappush(scoville, min1 + min2 * 2)
            answer += 1
    # 스코빌의 0번 째가 K보다 크면 정답
    if scoville[0] >= K:
        return answer
    # 아니면 나가
    else:
        return -1
    
    
    
